src/dosext/mfs/lfn.c

```diff
@@ -1342,3 +1342,75 @@ int mfs_lfn(void)
 		CARRY;
 	return ret;
 }
+
+static int mfs_remap_extended_open_(void)
+{
+  char *src = MK_FP32(_DS, _SI);
+  uint8_t dosint;
+  int drive, found;
+  char fpath[PATH_MAX];
+  struct stat st;
+
+  d_printf("SFN: extended open '%s'\n", src);
+
+  drive = build_posix_path(fpath, src, 0);
+  if (drive < 0)
+    return 0;
+
+  found = find_file(fpath, &st, drive, NULL);
+  d_printf("SFN: extended open '%s' does %sexist\n", fpath, found ? "" : "not ");
+
+  if ((_DX & 0b00000011) && !found)   // fail if doesn't exist
+    return lfn_error(FILE_NOT_FOUND); // 0b01 open // 0b10 truncate
+
+  if ((_DX & 0b00010000) && found)    // fail create if exists
+    return lfn_error(FILE_ALREADY_EXISTS);
+
+  dosint = (_AL == 1) ? 0x3c : 0x3d;  // select create / open function
+
+  /* Map the extended open into something the underlying DOS can understand */
+  /*
+     low byte of BX (only bit 2 lost)
+       bit 2 read-only no update access time
+
+     high byte of BX (all bits lost)
+       bit 8 disable buffering
+       bit 9 disable file compression
+       bit 10 use _DI value for file name mangling
+       bit 13 return error instead of int 24
+       bit 14 synchronous writes
+   */
+  _AL = (_BL & 0b11110011);
+
+  /* Filename passed in DS:DX */
+  _DX = _SI;
+
+  call_dos_helper(dosint);
+
+  /* Tell the caller what action was taken
+   *   (According to RBIL these values are lost when open uses redirector)
+     0x0001 - file opened
+     0x0002 - file created
+     0x0003 - file truncated
+   */
+  if (!isset_CF()) {
+    if (dosint == 0x3c)
+      _CX = found ? 0x0003 : 0x0002;
+    else
+      _CX = 0x0001;
+  }
+
+  return 1;
+}
+
+int mfs_remap_extended_open(void)
+{
+  int carry, ret;
+
+  carry = isset_CF();
+  ret = mfs_remap_extended_open_();
+  /* preserve carry if we forward the request */
+  if (ret == 0 && carry)
+    CARRY;
```

Comment:
You seem to have the same code at
a caller side:
```
+	if (!ret)
+	    CARRY;
```
which should be enough.

Reply:
I copied the code for mfs_lfn() is that wrong?

Reply:
No, you were actually right.
I reverted my patch that was not right.

Reply:
Please rebase.